### PR TITLE
[DRAFT][onert-micro] PoC_V1: Training for onert-micro

### PR DESCRIPTION
**Pull request by BalyshevArtem**
This is draft first version of training for onert-micro. Similar to https://github.com/Samsung/ONE/issues/12873#issuecomment-2059003990 proposal. 
Note: in this draft I added gradient related operation into main circle schema, not into circle+.
This draft also adds the necessary infrastructure to create a back propagation graph: circle-weight-divider and circle_training_configure tools (note: in https://github.com/Samsung/ONE/issues/12873#issuecomment-2059003990 proposal these two tools merged into one `TrainingConfigureTool`).

### CODE REVIEW

File: onert-micro/eval-driver/TrainingDriver.cpp [hunk 1]
```diff
@@ -100,118 +93,264 @@ int entry(int argc, char **argv)
...
+
+#if 0
+        calculateCrossEntropy(reinterpret_cast<float *>(train_interpreter.getOutputDataAt(0)),
+                              reinterpret_cast<float *>(cur_train_target_data),
+                              target_size);
+        printPredAndTargetsValues(reinterpret_cast<float *>(train_interpreter.getOutputDataAt(0)),
+                                  reinterpret_cast<float *>(cur_train_target_data),
+                                  target_size);
+#endif
+        train_interpreter.backward();
```
Comment (chunseoklee) [inline onert-micro/eval-driver/TrainingDriver.cpp:312]: @BalyshevArtem This approach (backward for each sample in a batch) does not produce the same result with backward mean/sum over batch approach ?

Comment (BalyshevArtem) [reply onert-micro/eval-driver/TrainingDriver.cpp:312]: I'm not sure I understood the question correctly. In this approach, for each current data example, we calculate the gradient and summarize it for all examples in the current batch sample. And then, using the `updateWeights` method, according to the chosen optimization technique (SGD, ADAM), we update the weights using the calculated gradients on the batch_size sample. Thus, we somehow take into account the gradients calculated on the entire batch_size sample

Comment (chunseoklee) [reply onert-micro/eval-driver/TrainingDriver.cpp:312]: >we calculate the gradient and summarize it for all examples in the current batch sample.

Oh, I will look into more. Could you please point to where we summerize the loss ?

Comment (BalyshevArtem) [reply onert-micro/eval-driver/TrainingDriver.cpp:312]: https://github.com/Samsung/ONE/blob/ad3942d79371e27e3534f3274a913605a75a1f78/onert-micro/onert-micro/src/core/train/OMTrainingRuntimeModule.cpp#L448-L454 - here summarize calculated gradients (SGD).
https://github.com/Samsung/ONE/blob/ad3942d79371e27e3534f3274a913605a75a1f78/onert-micro/onert-micro/src/core/train/OMTrainingRuntimeModule.cpp#L471-L473 - here updated weigths

File: onert-micro/eval-driver/TrainingDriver.cpp [hunk 2]
```diff
@@ -100,118 +93,264 @@ int entry(int argc, char **argv)
...
-                            reinterpret_cast<const uint8_t *>(label_train_data),
-                            reinterpret_cast<void *>(&mae_result));
+  // Test data
+  std::unique_ptr<float []> test_input_data(new float [input_size]);
+  std::unique_ptr<float []> test_target_data(new float [target_size]);
 
-  printf("MSE_ERROR TRAIN = %f\n", mse_result);
+  for (uint32_t e = 0; e < training_epochs; ++e)
+  {
+    train_interpreter.set_training_mode(true);
```
Comment (jyoungyun) [inline onert-micro/eval-driver/TrainingDriver.cpp:268]: @BalyshevArtem 

Does this code support transfer learning?

Comment (BalyshevArtem) [reply onert-micro/eval-driver/TrainingDriver.cpp:268]: Do you mean by transfer learning that only certain last layers are trained? If yes, then TrainingConfigure tool is responsible for how many and which specific layers are trained in this draft for `proposal_1` from https://github.com/Samsung/ONE/issues/12873#issuecomment-2059003990. 
from TrainingDriver we cannot change and choose which layers will be trained in this variant.

Comment (jyoungyun) [reply onert-micro/eval-driver/TrainingDriver.cpp:268]: I understood that the backprop graph contains which layers will be trained, and TrainingDriver cannot change the training layer information other than turning training on and off. I'm trying to add `trainable` property in circle+ file format. So I wonder how you handle the trasfer learning feature. Thank you for your explanation.
